tests/crypto/README.rst

```diff
@@ -0,0 +1,306 @@
++--------------------+-------------+----------------------------------------------------------------------------+----------------------------------------------------------------------------+
+
+Requirements
+************
+
+* One of the following development boards:
+
+  * |nRF9160DK|
+  * |nRF52840DK|
+  * |nRF52DK|
+  * |nRF51DK|
+
+.. note::
+   Nordic devices with smaller flash capacity, such as nRF51, nRF52810, or nRF52811 might not fit the full test suite.
+   A recommended approach in such case is to run subsets of the tests one by one.
+
+
+Building and running
+********************
+
+.. |test path| replace:: :file:`tests/crypto/`
+
+.. include:: /includes/build_and_run_test.txt
+
+There are multiple ways to build the tests.
+See :ref:`nrf_security_readme` for additional information about nrf_security configuration.
+You can use the following configuration files to build it in a specific setup:
+
+* :file:`overlay-backend-sw.conf` - software only, except a hardware module to generate entropy for RNG
+* :file:`overlay-backend-hw.conf` - hardware-accelerated using the Arm CryptoCell accelerator (for cryptography and entropy for RNG)
+* :file:`overlay-backend-hw-sw.conf` - a combination of hardware acceleration using Arm CryptoCell and a software implementation that adds key sizes and algorithms not supported in CryptoCell (the setup uses hardware as much as possible)
+
+.. _crypto_test_ztest_custom:
+
+Ztest custom log formatting
+===========================
+
+Cryptography tests replace the standard Ztest formatting to assure more efficient reporting of running tests and test results.
+Set the configuration option :option:`CONFIG_ZTEST_TC_UTIL_USER_OVERRIDE` to replace the Ztest macros ``TC_START`` and ``Z_TC_END_RESULT`` with versions more suited for reporting results of cryptographic tests.
+
+:option:`CONFIG_ZTEST_TC_UTIL_USER_OVERRIDE` uses :file:`./include_override/tc_util_user_override.h` to define the custom formatting.
+
+.. _crypto_test_testing:
+
+Testing
+=======
+
+1. Compile and program the application.
+#. Observe the result of the different test vectors in the log using `RTT Viewer`_ or a terminal emulator.
+   The last line of the output indicates the test result::
+
+      PROJECT EXECUTION SUCCESSFUL
+
+
+Additional test cases and test vectors
+======================================
+
+Test cases and test vectors can be added to the test suite either by including additional source files or by extending the existing files.
+
+
+Test case
+---------
+
+A test case is a function designed to verify parts of the functionality of a cryptographic operation.
+Most cryptographic operations, like hash calculations and ECDH, have multiple test cases to be able to cover all features.
+A typical test case is called by looping over the registered test vectors and calling the test case.
+The execution logs the verdict for each test vector.
+
+An example of an HMAC test case in a simplified form is shown below::
+
```

Comment:
I would suggest adding `hmac_setup` here as well, which does three things:

- clears buffers s.t. state from previous test vectors isn't carried over
- gets a pointer to the next test vector via the macro `ITEM_GET`. `ITEM_GET` takes three args (hmac example follows):
    - test_vector_hmac_data: The section to fetch a test vector from, should correspond to the vector registration sector name
    - test_vector_hmac_t: Which type of test vector is expected to be found in the section given
    - hmac_vector_n: Which test vector to fetch (0 indexed)

And also what `hmac_teardown` does, which is simply to increase the `hmac_vector_n` counter s.t. the next iteration fetches the next test vector.

Reply:
Can you come up with an alternate implemention so we can update this snippet?

Reply:
@torsteingrindvik will come back with a new snippet after finishing some other work

Reply:
This comment led me to change the 'Test case' section and divide it into 3 subsections:
- Registering a test case - this was a bit further down but I think it should be first?
- Setting up a test case - based on info I got from @torsteingrindvik 
- Executing a test case - what we have here at the place of this comment

Let me know if that is fine @frkv

Reply:
yes this is fine